### ISSUE TITLE
g-keys commands run on a loop

### ISSUE THREAD

**Issue by fulvouscoder** [mortendynamite/g19daemon#9]
Hello,

first and foremost, many thank you for this wonderful contribution.

I noticed a problem, whenever I press the G-Keys, I see the following events in the terminal going on a loop:

Run command:  ""
Run command:  ""
Run command:  ""
(...)

Notes:
I am running Manjaro 20 KDE with kernel 5.6.19-2;
None of the G-Keys are configured to execute a command;
I tested all keys in the keyboard, this only occurs when I press a G-Key.

Also tested this scenario:
I configured the G4 Key to open firefox using the following command: 

/usr/lib/firefox/firefox %u

When I press the G4 key, firefox is opened multiple times (I stopped counting at the 200 window)

Is this a known issue?
Is it possible I have something misconfigured, if so can you provide a clue?

Thank you for your assistance

**Comment by jimmyd-be** [comment]
I tried this with my installation (Kubuntu 20.04) but cannot reproduce this. Did you installed the daemon using the latest RPM file or did you compile the code from latest commit?

**Comment by jimmyd-be** [comment]
I will try to setup a Manjaro environment to try to reproduce this. It will take some time because I'm on holiday, so please be patience.

**Comment by fulvouscoder** [comment]
Hello @jimmyd-be 

Thanks for your work on this project!

I compiled the code based on the source.
I can redo it and provide any logs or files.
If you have any hints on where I would begin to look at, I can also put the time into the grind.

Let me know if I can assist in any way. :smile: 
Have a good  rest and enjoy your holidays!

Thanks

**Comment by jimmyd-be** [comment]
Currently it looks like the application receives multiple keypresses. I had the same issue but only I got 2 presses when I pressed my G key. I fixed that in G19daemon::gKeys(). Probably you receive multiple key actions for one actual keypress.

**Comment by jimmyd-be** [comment]
Maybe I just found a correct way to handle this. Now the code only receives press actions. So I remove the second action. But it should be better to run the code once until it receive an unpress event. I'm trying to implement this at the moment so maybe the fix can be created this week. I will make a different branch for this issue so that you can test if I fixed it.

**Comment by fulvouscoder** [comment]
Awesome!

I've been digging around the files and have the following points:

a) The menu titles of the plugins still have "Hello World"

keybacklight.cpp
`
screen = new Gscreen(icon, tr("Hello World"));
`

b) I noticed 2 different .conf files. the one over at Dynamite seems to save the settings that I change while using the keyboard menu keys.
However, after a system reboot, these configurations are not being loaded. I need to use the keyboard menu keys and enter the Keybacklight menu and press the OK key so that the configuration is again loaded.


/home/username/.config/Dynamite/G19Daemon.conf
`
[General]
Backlight=55
KeyBacklight_Blue=200
KeyBacklight_Green=255
KeyBacklight_Red=60
`
vs

/home/username/.config/G19Daemon/G19Daemon.conf
`
[General]
ActivePlugin=Clock
Backlight=255
m1_G1=
m1_G10=
m1_G11=
m1_G12=
m1_G2=
m1_G3=
m1_G4=
m1_G5=
m1_G6=
m1_G7=
m1_G8=
m1_G9=
m2_G1=
m2_G10=
m2_G11=
m2_G12=
m2_G2=
m2_G3=
m2_G4=
m2_G5=
m2_G6=
m2_G7=
m2_G8=
m2_G9=
m3_G1=
m3_G10=
m3_G11=
m3_G12=
m3_G2=
m3_G3=
m3_G4=
m3_G5=
m3_G6=
m3_G7=
m3_G8=
m3_G9=
`

Thanks for your assistance!

**Comment by jimmyd-be** [comment]
I will also change this. Thanks for investigating.

**Comment by fulvouscoder** [comment]
No problem!

I have another point:

c) After I setup my RGB using the LCD Keys (R=60,G=255,B=200), I press OK key and the configuration is saved to the Dynamite folder:
`
/home/username/.config/Dynamite/G19Daemon.conf
`
As soon as I use the Volume adjustment roller to change the sound volume, I notice that the keyboard lighting reacts to the change, going on a scale and when the sound is muted all the keyboard glows red. This is a very cool effect :smile: !!!

Now here lies the problem, after using the roller (or the Mute key), the keyboard lighting reacts accordingly, however it returns to a default state, which is not what I configured (the default seems to be a purpleish color).
It seems to me that the configuration file saved under Dynamite is only being loaded when the user saves changes with the LCD keys.

I am trying to find the piece of code that reads from that file and attempt to set it as the default, if it exists.

Hope I am not being too much work, but I just got excited with this project.
Thanks!

**Comment by jimmyd-be** [comment]
Is it possible to create a new ticket for this issue?
> Now here lies the problem, after using the roller (or the Mute key), the keyboard lighting reacts accordingly, however it returns to a default state, which is not what I configured (the default seems to be a purpleish color).
It seems to me that the configuration file saved under Dynamite is only being loaded when the user saves changes with the LCD keys.

I will do the rest of the points in this one. But it easier to follow progress if the large issues are placed in another ticket.

**Comment by fulvouscoder** [comment]
Of course!

I'll copy paste it to a new ticket.

Btw I was wondering if there is a debug mode, how are you doing it?

**Comment by jimmyd-be** [comment]
The main code can be debugged in QT creator or any other C++ ide. For debugging the plugins I use  `qDebug() << "test";`

**Comment by fulvouscoder** [comment]
Sounds like a plan.
Gonna grind it, wait for me on the otherside!

**Comment by jimmyd-be** [comment]
I have any progress on the looping. I'm testing it now on my computer and will commit it in a custom branch to test it.

**Comment by jimmyd-be** [comment]
@fulvouscoder can you please test changes on branch issue9 for the looping issue?

**Comment by fulvouscoder** [comment]
On it!
Just to be sure, there is no Make Uninstall right?

I have to manually delete the following files/folders:

`/home/username/.config/G19Daemon/G19Daemon.conf`

`/home/username/.config/Dynamite/G19Daemon.conf`

Taken from install_manifest.txt:
`
/usr/bin/g19daemon
/usr/share/applications/g19daemon.desktop
/etc/udev/rules.d/80-hid-g19.rules
/usr/lib/g19daemon/0.1/plugins/libhelloworld.so
/usr/lib/g19daemon/0.1/plugins/libkeybacklight.so
/usr/lib/g19daemon/0.1/plugins/libbacklight.so
/usr/lib/g19daemon/0.1/plugins/libpavolume.so
/usr/lib/g19daemon/0.1/plugins/libmpris.so
/usr/lib/g19daemon/0.1/plugins/libclock.so
/usr/lib/g19daemon/0.1/plugins/libhardwaremonitor.so
`

Am I missing something?

**Comment by jimmyd-be** [comment]
You don't need to uninstalll it when recompile the code. It will replace all the files.

I don't know a method to automaticly uninstalll the application.

**Comment by fulvouscoder** [comment]
I have an incoming wall of text... regarding the tests I've performed for your branch.

TL:DR

whenever I use a G Key, the event is detected and the command is executed, however the app hangs and I cannot use the LCD keys, switch M keys or any G Keys. I need to reboot the app.
In short, I can only use G Keys once and then the app hangs.

- Start app, followed by press of G1 Key with vlc command
$ g19daemon
Load Settings
Loading settings
Loading General settings
"Total screens loaded from settings: 0 of 0"
Loading sub screen order
Loading main screen order
"Loaded screen orders: 0"
Load StartScreen
Plugin initialized
"G1"  Pressed, Run command:  "/usr/bin/vlc %U"
VLC media player 3.0.11 Vetinari (revision 3.0.11-0-gdc0c5ced72)

- Start app, followed by press of G7 Key with '6' command
$ g19daemon
Load Settings
Loading settings
Loading General settings
"Total screens loaded from settings: 0 of 0"
Loading sub screen order
Loading main screen order
"Loaded screen orders: 0"
Load StartScreen
Plugin initialized
"G7"  Pressed, Run command:  "6"

**Comment by jimmyd-be** [comment]
The text is normal and is showing the debug logs on startup. The hanging isnt good

**Comment by fulvouscoder** [comment]
# Tests performed and results:

- Tested keyboard keys - **No issues**
- Tested LCD keys - **No issues**
- Tested M Keys - **N/A** (No issues: keys change and event is detected; However not sure how to test MR Key functionality)
- Tested G Keys - **Issue**, can only use G Keys once, after that app hangs
- Volume adjustment roller - **No issues** (keyboard lightning reacts to volume changes, mute status and returns to previously configured Keybacklight settings a.k.a. G19Daemon.conf values)
- Logitech G19 Daemon (gui) - **No issues**

I am adding the detailed test information, but that is the general state of affairs.

On another good note, it seems you also fixed the issue with the Keybacklight, everytime I restart the app, the settings are being loaded and I don't see that Dynamite folder.

**Comment by jimmyd-be** [comment]
Tnx for testing the full daemon.

Currently there aren't any actions binded to the MR-key. If you have any idea what can be handy to link to that button please suggest. This afternoon I will investigate application hanging after using G-keys.

**Comment by fulvouscoder** [comment]
No problem!

Let me know if you need to have some kind of specific test done or any other that I missed.

I'll think about the MR key. right now I am not even sure how to properly use the G Keys.

Do you happen to use macros, assigned to G keys?
For example, G3 = Key Press 'Shift' + Key Press 'K' + Key Release 'K' + Key Press 'O' + Key Release 'O' + Key Release 'Shift'
- Pressing G3 will execute the "Hold Shift to type KO"

I am also looking into how to do this for the first time.

**Comment by jimmyd-be** [comment]
Currently u only use the g Keys to open applications. Maybe you van try to simulate keystrokes using Shell scripts and run the script using the g Keys. Or we can add this as a new feature to simulate special keystrokes. But then please create a ticket for this and I will investigate this.

**Comment by jimmyd-be** [comment]
Yesteday I tried to find the application hanging but couldn't found it. On Kubuntu 20.04 it is working. I need to try to run manjaro to find the issue but it will take some time to debug in Manjaro to fix this.

**Comment by fulvouscoder** [comment]
no worries!

Is there a way to restart the app?
If the app is running and you send the command to start the daemon, the message "App is running" pops up:

$ g19daemon 
App is running

however in some instances during my grind, I've noticed there is no daemon running, however you still get the same message:

$ g19daemon 
App is running
$ ps -ef | grep g19
username    22404    1557  0 14:16 pts/1    00:00:00 grep --colour=auto g19

I was wondering if it's some variable, linked to the hardware, that is "turned on".
Have you ever encountered this?

I am still testing the waters with CT Creator.
- I have imported the project and while looking at some code in:
Source Files -> g19daemon.cpp
- performed the action
[Debug] -> [Start debugging without deployment]

Am I bringing too much dialogue into this thread? :nerd_face:

**Comment by jimmyd-be** [comment]
This is because the deamon has not safly closed. Look in main.cpp file for the name of the file that is placed when the app is starting and need to be removed manually when you close the app without using the close button.

**Comment by fulvouscoder** [comment]
Thanks, I found the file: 

`$ ls -lsrth /run/user/1000/g19deamon.pid`

I see a warning for this line, could it pose a problem?

`g19daemon.cpp` => `void G19daemon::gKeys()` =>  `QProcess::startDetached(command);`

void G19daemon::gKeys() {
(...)

  if (!gKey.isEmpty()) {
      QString command =
          settings->value(translateKey(device->getActiveMKey()) + "_" + gKey)
              .toString();
      qDebug() << gKey << " Pressed, Run command: " << command;
      QProcess::startDetached(command);
  }
(...)
}

- This is the warning:
"
g19daemon.cpp:196:17: warning: 'startDetached' is deprecated: Use QProcess::startDetached(const QString &program, const QStringList &arguments) instead
qprocess.h:275:5: note: 'startDetached' has been explicitly marked deprecated here
qglobal.h:294:33: note: expanded from macro 'QT_DEPRECATED_X'
qcompilerdetection.h:675:55: note: expanded from macro 'Q_DECL_DEPRECATED_X'
"

**Comment by jimmyd-be** [comment]
Can you try to comment QProcess::startDetached(command) and try to run the application. If the deamon is not hanging anymore then we found the hanging issue.

**Comment by fulvouscoder** [comment]
No more warnings, but it still hangs.

`//QProcess::startDetached(command);`
"
16:20:12: Starting /home/username/GitRepos/build-g19daemon-issue9-Desktop-Debug/g19daemon ...
"m2"  Pressed, Run command:  ""
"m3"  Pressed, Run command:  ""
"m1"  Pressed, Run command:  ""
"m3"  Pressed, Run command:  ""
"m1"  Pressed, Run command:  ""
"G7"  Pressed, Run command:  "5"
"

**Comment by jimmyd-be** [comment]
Maybe we need to look into usblib. I thing the problem is sitting here. For example
https://stackoverflow.com/q/39861085/2448399

**Comment by jimmyd-be** [comment]
I have installed Manjaro to test it from there. After building and installing the G19Daemon I configured the M1-G1 key to open Konsole. I tried that and it opens Konsole. After pressing another G-key it shows the key has been pressed and every time I press the G1 key the Konsole application is shown.

I installed Manjaro Kde 2020.0.3 with kernel 5.4.52. So I think the problem is something on your setup. An application that is not compatible or some settings that breaks the G19daemon. Currently I cannot fix this because I cannot reproduce it.

**Comment by fulvouscoder** [comment]
Thanks for running that test!

It's weird, even if I don't have anything assigned to the G key, same thing happens.

I am going to create a sandbox with manjaro and retrace the steps. Thinking about software in my daily driver, If I had to make a guess ( an uneducated one ) I do have Piper/libratbag installed.

I'll get back to you asap.

**Comment by jimmyd-be** [comment]
I ran the command `sudo dmesg | grep usb` and I also have

> [   25.379767] usb 3-9.1.1.2: usbfs: process 2734 (g19daemon) did not claim interface 0 before use

**Comment by fulvouscoder** [comment]
Thanks!
I will get back at you, as soon as I test with a clean environment. Hopefully this weekend :crossed_fingers:
